pkg/ccl/backupccl/testdata/backup-restore/procedures

```diff
@@ -0,0 +1,393 @@
+# Test backing up and restoring a database with procedures.
+new-cluster name=s
+----
+
+exec-sql
+CREATE DATABASE db1;
+USE db1;
+CREATE SCHEMA sc1;
+CREATE TABLE sc1.tbl1(a INT PRIMARY KEY);
+CREATE TYPE sc1.enum1 AS ENUM('Good');
+CREATE SEQUENCE sc1.sq1;
+CREATE PROCEDURE sc1.p1(a sc1.enum1) LANGUAGE SQL AS $$
+  SELECT a FROM sc1.tbl1;
+  SELECT nextval('sc1.sq1');
+$$;
+CREATE SCHEMA sc2;
+CREATE TABLE sc2.tbl2(a INT PRIMARY KEY);
+CREATE PROCEDURE sc2.p2(i INT) LANGUAGE SQL AS $$
+  INSERT INTO sc2.tbl2 VALUES (i)
+$$;
+----
+
+exec-sql
+CALL sc2.p2(123)
+----
+
+query-sql
+SELECT * FROM sc2.tbl2
+----
+123
+
+exec-sql
+BACKUP DATABASE db1 INTO 'nodelocal://1/test/'
+----
+
+query-sql
+WITH descs AS (
+  SHOW BACKUP LATEST IN 'nodelocal://1/test/'
+)
+SELECT database_name, parent_schema_name, object_name, object_type, is_full_cluster FROM descs
+----
+<nil> <nil> db1 database false
+db1 <nil> public schema false
+db1 <nil> sc1 schema false
+db1 sc1 tbl1 table false
+db1 sc1 enum1 type false
+db1 sc1 _enum1 type false
+db1 sc1 sq1 table false
+db1 sc1 p1 function false
+db1 <nil> sc2 schema false
+db1 sc2 tbl2 table false
+db1 sc2 p2 function false
+
+query-sql
+SELECT create_statement FROM [SHOW CREATE PROCEDURE sc1.p1]
+----
+CREATE PROCEDURE sc1.p1(IN a db1.sc1.enum1)
+	LANGUAGE SQL
+	AS $$
+	SELECT a FROM db1.sc1.tbl1;
+	SELECT nextval('sc1.sq1'::REGCLASS);
+$$
+
+exec-sql
+CALL sc1.p1('Good'::sc1.enum1)
+----
+
+query-sql
+SELECT currval('sc1.sq1')
+----
+1
+
+exec-sql
+DROP DATABASE db1
+----
+
+exec-sql
+RESTORE DATABASE db1 FROM LATEST IN 'nodelocal://1/test/' WITH new_db_name = db1_new
+----
+
+exec-sql
+USE db1_new
+----
+
+# Make sure procedure ids in signature and body are rewritten.
+# 1. argument type id is rewritten so that type name is deserialized correctly.
+# 2. db name in qualified name is rewritten.
+# 3. sequence id is rewritten so that sequence name is deserialized correctly.
+query-sql
+SELECT create_statement FROM [SHOW CREATE PROCEDURE sc1.p1]
+----
+CREATE PROCEDURE sc1.p1(IN a db1_new.sc1.enum1)
+	LANGUAGE SQL
+	AS $$
+	SELECT a FROM db1_new.sc1.tbl1;
+	SELECT nextval('sc1.sq1'::REGCLASS);
+$$
+
+# Make sure procedure signature is rewritten in schema descriptor so that
+# procedure can be resolved and executed.
+exec-sql
+CALL sc1.p1('Good'::sc1.enum1)
+----
+
+query-sql
+SELECT currval('sc1.sq1')
+----
+1
+
+# Make sure procedure still inserts into the correct table.
+exec-sql
+CALL sc2.p2(456)
+----
+
+query-sql
+SELECT * FROM sc2.tbl2
+----
+123
+456
+
+# Make sure dependency IDs are rewritten.
+# Note that technically this only tests forward-reference IDs in depended-on
+# objects are rewritten. But since we have cross-references validation, so this
+# also means back-references in UDF descriptor are good.
```

Comment:
nit: s/UDF/StoredProcedure/r? I see a few other examples in the test.

Reply:
Technically, they are function descriptors—there aren't separate descriptors for procedures and UDFs, they are shared. But I've reworded it a bit to avoid UDF since that's not completely accurate either.